wrappers/python/nntile/layer/llama_attention.py

```diff
@@ -893,6 +904,410 @@ def forward_async(self):
+            + tuple(k.basetile_shape[2:]),
+        )  # (n_seq, n_seq, batch=n_batch, batch=n_head)
+        a_maxsumexp_tmp = nntc.empty(
+            (2,) + tuple(a_tmp.shape[1:]),
+            dtype=type(q),
+            basetile_shape=(2,) + tuple(a_tmp.basetile_shape[1:]),
+        )
+        b_tmp = nntc.empty(
+            q.shape,
+            dtype=type(q),
+            basetile_shape=q.basetile_shape,
+        )  # (head_size, n_seq, n_batch, n_head)
+        b_tr_tmp = nntc.empty(
+            tuple(b_tmp.shape[3:]) + tuple(b_tmp.shape[:3]),
+            dtype=type(q),
+            basetile_shape=tuple(b_tmp.basetile_shape[3:])
+            + tuple(b_tmp.basetile_shape[:3]),
+        )  # (n_head, head_size, n_seq, n_batch)
+
+        y_tensor = nntc.empty(
+            (self.w.value.shape[0],) + tuple(q.shape[1:3]),
+            dtype=type(q),
+            basetile_shape=(self.w.value.basetile_shape[0],)
+            + tuple(q.basetile_shape[1:3]),
+        )
+
+        gemm_async(
+            1.0 / self.head_size**0.5,
+            trans,
+            k,
+            notrans,
+            q,
+            0.0,
+            a_tmp,
+            1,
+            3,
+            redux=self.redux,
+        )
+        clear_async(a_maxsumexp_tmp)
+        if self.mask:
+            mask_tmp = nntc.empty(a_tmp.shape[:2], dtype=Tensor_bool)
+            copy_intersection_async(
+                self.mask, [0, 0], mask_tmp, [0, k.shape[1] - q.shape[1]]
+            )
+            mask_scalar_async(mask_tmp, self.val, a_tmp, 3)
+
+        maxsumexp_async(a_tmp, a_maxsumexp_tmp, 0, redux=self.redux)
+        softmax_inplace_async(a_maxsumexp_tmp, 1.0, a_tmp, 0)
+
+        gemm_async(
+            1.0,
+            notrans,
+            v,
+            notrans,
+            a_tmp,
+            0.0,
+            b_tmp,
+            1,
+            3,
+            redux=self.redux,
+        )
+
+        transpose_async(1.0, b_tmp, b_tr_tmp, 3)
+
+        gemm_async(
+            1.0,
+            notrans,
+            self.w.value,
+            notrans,
+            b_tr_tmp,
+            0.0,
+            y_tensor,
+            3,
+            0,
+            redux=self.redux,
+        )
+
+        if self.out_proj_bias is not None:
+            add_fiber_async(1.0, self.out_proj_bias.value, 1.0, y_tensor, 0, 0)
+
+        return y_tensor
+
+    def _apply_rope_dynamic(
+        self, x: Tensor, q_partial: Tensor, k_partial: Tensor
+    ):
+        q_rope_partial = nntc.empty(
+            q_partial.shape,
+            basetile_shape=q_partial.basetile_shape,
+            dtype=type(x),
+        )
+        k_rope_partial = nntc.empty(
+            k_partial.shape,
+            basetile_shape=k_partial.basetile_shape,
+            dtype=type(x),
+        )
+
+        sin_partial = nntc.zeros((self.sin.shape[0],) + tuple(x.shape[-2:]))
+        cos_partial = nntc.zeros((self.cos.shape[0],) + tuple(x.shape[-2:]))
+        copy_intersection_async(
```

Comment:
We need to think how to rework this copying from `sin` and `cos`. Easy solution is to divide `sin` and `cos` into tiles of size 1 in a corresponding axis.

Reply:
there should be slice operation, working like view, which do virtual slice without actual copy
dividing into slices - seems to be very implementation dependent solution

Reply:
Unfortunately, supporting `view` is impossible with StarPU approach.

Reply:
oh. Probably need to think about it separately. Because I have already used this approach several times. And all this code need to be rethought